File: backend/app/schemas/stats.py
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field


class PricePerM2Stats(BaseModel):
    """Статистика цены за м²."""

    average: float = Field(..., description="Средняя цена за м²")
    median: float = Field(..., description="Медианная цена за м²")
    min: float = Field(..., description="Минимальная цена за м²")
    max: float = Field(..., description="Максимальная цена за м²")
    count: int = Field(..., description="Количество объявлений")
    currency: str = Field(..., description="Валюта (byn/usd)")


class PricePerM2Trend(BaseModel):
    """Тренд цены за м² по времени."""

    date: str = Field(..., description="Дата (YYYY-MM-DD)")
    average: float = Field(..., description="Средняя цена за м²")
    median: float = Field(..., description="Медианная цена за м²")
    count: int = Field(..., description="Количество объявлений")


class PricePerM2Distribution(BaseModel):
    """Распределение цены за м² (гистограмма)."""

    range_min: float = Field(..., description="Минимальная граница диапазона")
    range_max: float = Field(..., description="Максимальная граница диапазона")
    count: int = Field(..., description="Количество объявлений в диапазоне")
    percentage: float = Field(..., description="Процент от общего количества")


class PricePerM2StatsRequest(BaseModel):
    """Запрос статистики цены за м²."""

    city: str = Field(..., description="Город")
    rooms: Optional[int] = Field(None, ge=1, le=10, description="Количество комнат")
    date_from: Optional[datetime] = Field(None, description="Дата от")
    date_to: Optional[datetime] = Field(None, description="Дата до")
    currency: str = Field("usd", description="Валюта (byn/usd)")


class PricePerM2TrendsRequest(BaseModel):
    """Запрос трендов цены за м²."""

    city: str = Field(..., description="Город")
    rooms: Optional[int] = Field(None, ge=1, le=10, description="Количество комнат")
    period_days: int = Field(30, ge=1, le=365, description="Период в днях")
    interval: str = Field("day", description="Интервал группировки (day/week/month)")
    currency: str = Field("usd", description="Валюта (byn/usd)")


class PricePerM2DistributionRequest(BaseModel):
    """Запрос распределения цены за м²."""

    city: str = Field(..., description="Город")
    rooms: Optional[int] = Field(None, ge=1, le=10, description="Количество комнат")
    bins: int = Field(10, ge=5, le=50, description="Количество бинов")
    currency: str = Field("usd", description="Валюта (byn/usd)")

File: frontend/src/shared/types/stats.ts
/**
 * Типы для аналитики цены за м²
 */

export interface PricePerM2Stats {
  average: number;
  median: number;
  min: number;
  max: number;
  count: number;
  currency: 'byn' | 'usd';
}

export interface PricePerM2Trend {
  date: string;
  average: number;
  median: number;
  count: number;
}

export interface PricePerM2TrendsResponse {
  city: string;
  period_days: number;
  interval: 'day' | 'week' | 'month';
  currency: 'byn' | 'usd';
  data: PricePerM2Trend[];
}

export interface PricePerM2DistributionBin {
  range_min: number;
  range_max: number;
  count: number;
  percentage: number;
}

export interface PricePerM2DistributionResponse {
  city: string;
  bins: number;
  currency: 'byn' | 'usd';
  data: PricePerM2DistributionBin[];
}

/**
 * Типы для Deal Finder (поиск выгодных квартир)
 */

export interface DealListing {
  id: string;
  kufar_id: string;
  url: string;
  title: string;
  price: number;
  price_usd?: number | null;
  currency: string;
  city: string | null;
  address: string | null;
  rooms: number | null;
  area: number | null;
  floor: number | null;
  total_floors?: number | null;
  // ✅ Дополнительные поля для совместимости с Listing (точно такие же типы)
  category: string | null;
  description: string | null;
  district: string | null;
  metro: string | null;
  house_year: number | null;
  images: string[];
  status: string;
  first_seen_at: string;
  last_seen_at: string;
  price_per_m2_byn?: number | null;
  price_per_m2_usd?: number | null;
  deal_percent: number;
  avg_price_per_m2: number;
}

export interface DealsResponse {
  items: DealListing[];
  total: number;
  limit: number;
  offset: number;
  avg_price_per_m2: number;
  currency: string;
}

export interface DealStats {
  total_deals: number;
  avg_deal_percent: number;
  best_deal_percent: number;
  avg_price_per_m2: number;
  total_savings: number;
}

/**
 * Типы для Price Drop Tracker (трекинг падения цены)
 */

export interface PriceDropListing {
  id: string;
  kufar_id: string;
  url: string;
  title: string;
  price: number;
  price_usd?: number | null;
  currency: string;
  city: string | null;
  address: string | null;
  rooms: number | null;
  area: number | null;
  floor: number | null;
  total_floors?: number | null;
  category: string | null;
  description: string | null;
  district: string | null;
  metro: string | null;
  house_year: number | null;
  images: string[];
  status: string;
  first_seen_at: string;
  last_seen_at: string;
  price_per_m2_byn?: number | null;
  price_per_m2_usd?: number | null;
  // Price drop метрики
  max_price: number;
  min_price: number;
  drop_percent: number;
  current_price: number;
}

export interface PriceDropResponse {
  items: PriceDropListing[];
  total: number;
  avg_drop_percent: number;
  max_drop_percent: number;
  min_drop_percent: number;
  currency: string;
}

export interface PriceDropHistoryItem {
  event_type: string;
  price_before: number;
  price_after: number;
  price_before_usd?: number | null;
  price_after_usd?: number | null;
  created_at: string;
}

export interface PriceDropHistoryResponse {
  items: PriceDropHistoryItem[];
  first_price: number;
  last_price: number;
  total_drop_percent: number;
}

/**
 * Типы для Deal Score (оценка выгодности объявления)
 */

export interface DealScoreBreakdown {
  score: number;          // Score фактора (0-100)
  weight: number;         // Вес фактора
  weighted: number;       // Взвешенный вклад (score * weight / 100)
}

export interface DealScoreFullBreakdown {
  price_score: DealScoreBreakdown;      // Цена ниже рынка
  trend_score: DealScoreBreakdown;      // Падение цены
  liquidity_score: DealScoreBreakdown;  // Дней на рынке
  freshness_score: DealScoreBreakdown;  // Свежее объявление
  floor_score: DealScoreBreakdown;      // Этаж
  bonus_score: DealScoreBreakdown;      // Бонусы (фото, площадь)
}

export interface DealScoreListing {
  id: string;
  kufar_id: string;
  url: string;
  title: string;
  price: number;
  price_usd?: number | null;
  currency: string;
  city: string | null;
  address: string | null;
  rooms: number | null;
  area: number | null;
  floor: number | null;
  total_floors?: number | null;
  category: string | null;
  description: string | null;
  district: string | null;
  metro: string | null;
  house_year: number | null;
  images: string[];
  status: string;
  first_seen_at: string;
  last_seen_at: string;
  price_per_m2_byn?: number | null;
  price_per_m2_usd?: number | null;
  // Deal Score поля
  deal_score: number;           // 0-100
  deal_label: string;           // "🔥 HOT", "👍 GOOD", "😐 NORMAL"
  deal_score_breakdown?: DealScoreFullBreakdown;  // Детализация по факторам
}

export interface DealsScoreResponse {
  items: DealScoreListing[];
  total: number;
  avg_score: number;
  min_score_filter: number;
  currency: string;
  limit: number;
  offset: number;
}
